File: C#/House/House.cs
using System;
using System.Collections.Generic;
using System.Linq;

public record ChainItem(string Thing, string Action);

public static class House
{
    private static readonly List<ChainItem> chain = new List<ChainItem>
    {
        new ChainItem("the house that Jack built.", ""),
        new ChainItem("the malt", "lay in"),
        new ChainItem("the rat", "ate"),
        new ChainItem("the cat", "killed"),
        new ChainItem("the dog", "worried"),
        new ChainItem("the cow with the crumpled horn", "tossed"),
        new ChainItem("the maiden all forlorn", "milked"),
        new ChainItem("the man all tattered and torn", "kissed"),
        new ChainItem("the priest all shaven and shorn", "married"),
        new ChainItem("the rooster that crowed in the morn", "woke"),
        new ChainItem("the farmer sowing his corn", "kept"),
        new ChainItem("the horse and the hound and the horn", "belonged to")
    };

    public static string Recite(int verseNumber)
    {
        var lines = new List<string> { $"This is {chain[verseNumber - 1].Thing}" };

        for (int i = 1; i < verseNumber; i++)
        {
            lines.Add($"that {chain[verseNumber - i].Action} {chain[verseNumber - i - 1].Thing}");
        }

        return string.Join(" ", lines);
    }

    public static string Recite(int startVerse, int endVerse)
    {
        var lines = new List<string>();

        for (int i = startVerse; i <= endVerse; i++)
        {
            lines.Add(Recite(i));
        }

        return string.Join("\n", lines);
    }
}

File: JavaScript/house.js
class House {
    static chain = [
        { thing: 'the house that Jack built.', action: '' },
        { thing: 'the malt', action: 'lay in' },
        { thing: 'the rat', action: 'ate' },
        { thing: 'the cat', action: 'killed' },
        { thing: 'the dog', action: 'worried' },
        { thing: 'the cow with the crumpled horn', action: 'tossed' },
        { thing: 'the maiden all forlorn', action: 'milked' },
        { thing: 'the man all tattered and torn', action: 'kissed' },
        { thing: 'the priest all shaven and shorn', action: 'married' },
        { thing: 'the rooster that crowed in the morn', action: 'woke' },
        { thing: 'the farmer sowing his corn', action: 'kept' },
        { thing: 'the horse and the hound and the horn', action: 'belonged to' },
    ];

    static verse(id) {
        let lines = [`This is ${House.chain[id - 1].thing}`];
    
        for (let i = 1; i < id; i++) {
            lines.push(`that ${House.chain[id - i].action} ${House.chain[id - i - 1].thing}`);
        }
    
        return lines;
    }
  
    static verses(start, end) {
        let lyrics = [];
     
        for (let i = start; i <= end; i++) {
            lyrics = lyrics.concat(House.verse(i));
            lyrics = lyrics.concat('');
        }
     
        lyrics.pop();
     
        return lyrics;
    }
}

console.log(House.verses(4, 8));